### PR TITLE
SAML based SSO 

### LINKED ISSUE #4974: Improve SSO Integration

### PR DESCRIPTION
**Pull request by Asrani-Aman**
Fixes #4974 

#### Describe the changes you have made in this PR -
1. fix `devise_saml_authenticable` and `ruby-saml` gem versions compatibility
2. added `saml_update_resource_hook` (this resolves overriding  issue of create method in saml_session_controller by devise_authenticable_gem  )
3. added Routes for controller under` user `scope

### Screenshots of the changes (If any) -

https://github.com/CircuitVerse/CircuitVerse/assets/96644946/92754ccc-a7ce-4acd-8423-dc2ab36b6253

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -221,3 +220,6 @@ gem "maintenance_tasks", "~> 2.3"
 gem "stackprof"
 gem "sentry-ruby"
 gem "sentry-rails"
+# for SAML based SSO
+gem 'devise_saml_authenticatable', '~> 1.9.1'
+gem 'ruby-saml', '~> 1.16.0'
```
Comment (tachyons) [inline Gemfile:225]: You don't need to add this gem explicitly as it is dependency of devise_saml_authenticatable

File: config/initializers/devise.rb [hunk 2]
```diff
@@ -311,3 +311,21 @@
...
+    user.send "#{key}=", saml_response.attribute_value_by_resource_key(key)
+  end
+
+  if Devise.saml_use_subject
+    user.send "#{Devise.saml_default_user_key}=", auth_value
+  end
+
+  puts "User: #{user.inspect}"
+  if user.id.nil?
+    user.name = user.email
```
Comment (tachyons) [inline config/initializers/devise.rb:327]:
```suggestion
    user.name = user.email if user.name.blank?
```

File: config/initializers/devise.rb [hunk 3]
```diff
@@ -311,3 +311,21 @@
...
+  # Copied from the default behavior: https://github.com/apokalipto/devise_saml_authenticatable/blob/09221d59e434bc947327e478a388013a70add4f7/lib/devise_saml_authenticatable.rb#L101
+  saml_response.attributes.resource_keys.each do |key|
+    user.send "#{key}=", saml_response.attribute_value_by_resource_key(key)
+  end
+
+  if Devise.saml_use_subject
+    user.send "#{Devise.saml_default_user_key}=", auth_value
+  end
+
+  puts "User: #{user.inspect}"
```
Comment (smritigarg) [inline config/initializers/devise.rb:325]: **Question** Is this statement intentional?

**Comment by smritigarg** [comment]
@Asrani-Aman I  checked the video and basic implementation looks fine so far. Please address the review comments. In the ui I see `sign-on` button labled as `Single Sign-on` .. Its a bit misleading may be we can rename it to `Single Sign on with Okta` ( I suppose thats what we are supporting as of now ?). Also please commit `Gemfile.lock` and one `sample devise config` for Single sign on configurations

**Comment by Asrani-Aman** [comment]
@tachyons @smritigarg 

Good evening mentors,

I've removed the `ruby-saml gem`, and everything is now working smoothly without it.

and also as you suggest  I don't need to explicitly define the methods in `saml_session_controller.rb` as it might not handle all the cases and Instead, using the implicit Devise methods works well. The only thing we need to add is `Devise.saml_update_resource_hook`; otherwise, it will cause errors as mentioned in attached video in PR

**Comment by Asrani-Aman** [comment]
@smritigarg

I tested it with Okta, and I believe it will work with other IDPs as well. We just need to add the respective ['IDP_SSO_URL'] and ['IDP_FINGERPRINT'] of each Identity Provider in the .env file.

Could you please provide more context regarding `adding the Devise config for Single Sign-On` configurations? 
Do you mean adding documentation?

File: config/initializers/devise.rb (hunk 3 above)
Comment (Asrani-Aman) [reply config/initializers/devise.rb]: Sorry @smritigarg, I added this for debugging and forgot to remove it before committing the changes.

File: Gemfile [hunk 4]
```diff
@@ -221,3 +220,5 @@ gem "maintenance_tasks", "~> 2.3"
 gem "stackprof"
 gem "sentry-ruby"
 gem "sentry-rails"
+# for SAML based SSO
+gem 'devise_saml_authenticatable', '~> 1.9.1'
```
Comment (smritigarg) [inline Gemfile:224]: I do not think there is a need for two entries for `gem 'devise_saml_authenticatable'`, lets remove the entry without version here. Run bundle install again and also commit updated `Gemfile.lock`

**Comment by Asrani-Aman** [comment]
@smritigarg done Changes

**Comment by tachyons** [comment]
Could you add documentation for how to configure Circuitverse with SAML using Okta and keycloak ?

**Comment by tachyons** [comment]
Also update Pr description

**Comment by Asrani-Aman** [comment]
@tachyons @smritigarg 
- Added the Documentation for or how to configure Circuitverse with SAML using Okta and keycloak 

May you please give a look at it [PR Link](https://github.com/CircuitVerse/CircuitVerseDocs/pull/380)

**Pull request merged**